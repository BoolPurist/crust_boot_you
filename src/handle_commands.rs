use crate::{
    cli::SaveTemplateCli,
    file_management::{self, FileKind},
    prelude::*,
    AppCliEntry, NotEmptyText, SubCommands,
};
use std::path::Path;

pub type ReturnType = AppResult<String>;

pub fn handle(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    args: &AppCliEntry,
) -> ReturnType {
    match args.sub_commands() {
        SubCommands::LoadTemplate { name } => {
            handle_load_template(path_provider, file_manipulator, name)
        }
        SubCommands::SaveTemplate(args) => {
            handle_save_template(path_provider, file_manipulator, args)
        }
    }
}

fn handle_load_template(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    name: &NotEmptyText,
) -> ReturnType {
    debug!("Handling subcommand: {:?}", "LoadTemplate");

    let path_to_template = path_provider.specific_entry_template_files(name)?;

    if !path_to_template.exists() {
        bail!("No content for teamplate named {}, could be found", name);
    }

    let cwd = path_provider.cwd()?;
    file_manipulator.copy_dir(path_to_template, &cwd)?;

    Ok(format!(
        "Folder {:?} filled with content from Template ({})",
        cwd, name
    ))
}

fn save_template(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    name: &NotEmptyText,
    path: impl AsRef<Path>,
) -> ReturnType {
    debug!("Handling subcommand: {:?}", "SaveTemplate");
    let (source_path, template_name) = (path.as_ref(), name);
    info!("Copying from {:?}", source_path);
    let file_type = file_management::detect_file_kind(source_path)?;

    match file_type {
        FileKind::File => {
            debug!("Detected {:?} as file", source_path);
            let ensured_template_folder = file_management::ensure_target_template_folder(
                &path_provider,
                &file_manipulator,
                template_name,
            )?;

            let target_path =
                file_management::construct_file_target_path(source_path, &ensured_template_folder)?;
            debug!(
                "Copying from source path {:?} to target path {:?}",
                source_path, target_path,
            );
            file_manipulator.copy_file(source_path, target_path)?;
        }
        FileKind::Folder => {
            debug!("Detected {:?} as folder", source_path);
            let target_path = file_management::ensure_target_template_folder(
                &path_provider,
                &file_manipulator,
                template_name,
            )?;
            debug!(
                "Copying from source path {:?} to target path {:?}",
                source_path, target_path,
            );
            file_manipulator.copy_dir(source_path, target_path)?;
        }
    }

    let msg_to_user = format!(
        "Created template with name ({}) from the {:?}",
        name, source_path,
    );
    Ok(msg_to_user)
}
fn handle_save_template(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    from_cli: &SaveTemplateCli,
) -> ReturnType {
    let arguments = from_cli.arguments();
    let (name, path) = (arguments.name(), arguments.path());
    save_template(path_provider, file_manipulator, name, path)
}
