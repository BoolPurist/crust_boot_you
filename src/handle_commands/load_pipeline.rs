#![allow(dead_code)]
/// TODO:DEAD_CODE
use std::{borrow::Cow, collections::HashSet};

use crate::{
    cli::InitKind,
    file_management::{
        write_transactions::{DirToEnsure, FileToLoad, WriteTransactions},
        FileKind, NodeEntryMeta, SourceTargetNode,
    },
    prelude::*,
    template_augmentation::TemplateAugmentor,
};

type NodeId<'a> = (&'a Path, FileKind);
type AugmentedFile<'a> = (Cow<'a, str>, &'a Path);

#[derive(Debug)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
pub enum InitAction {
    NotEmpty,
    Conflict(PathBuf),
    NoConflict(Vec<SourceTargetNode>),
    Purge(Vec<SourceTargetNode>),
}

pub fn init_project_with_template(
    file_manipulator: &impl FileManipulator,
    augmentor: &mut impl TemplateAugmentor,
    init_kind: InitKind,
    write_target: &Path,
    template_source: &Path,
) -> AppResult<()> {
    let action = determine_init_action(init_kind, write_target, template_source, |path| {
        file_manipulator.all_nodes_inside(path)
    })?;
    return match action {
        InitAction::NotEmpty => bail!(
            "Aborted initialization of project. Reason: target location at {:?} is not empty. This is not allowed for init kind {}", 
            write_target, 
            init_kind
        ),
        InitAction::Conflict(conflict) => bail!(
            "Aborted initialization of project. Reason: found file/folder at {:?} prevents it for current init kind {}", 
            conflict, 
            init_kind
        ),
        InitAction::NoConflict(data) => {
            let (files, dirs) = WriteTransactions::new(data).into();
            let loaded = load_all_files(file_manipulator, files)?;
            let augmented = augment_loaded_files(augmentor, &loaded)?;
            ensure_folders(file_manipulator, &dirs)?;
            write_loaded_files(file_manipulator, &augmented)?;
            Ok(())
        }
        InitAction::Purge(data) => {
            let (files, dirs) = WriteTransactions::new(data).into();
            let loaded = load_all_files(file_manipulator, files)?;
            let augmented = augment_loaded_files(augmentor, &loaded)?;

            file_manipulator.delete_whole_folder(write_target)?;

            ensure_folders(file_manipulator, &dirs)?;
            write_loaded_files(file_manipulator, &augmented)?;
            Ok(())
        }
    };
}

fn write_loaded_files<'a>(
    file_manipulator: &impl FileManipulator,
    to_write: &[AugmentedFile<'a>],
) -> AppResult {
    for (content, path) in to_write {
        file_manipulator.write_file_to(path, content)?;
    }
    Ok(())
}

fn ensure_folders(
    file_manipulator: &impl FileManipulator,
    to_ensure: &[DirToEnsure],
) -> AppResult {
    for next in to_ensure {
        file_manipulator.ensure_dir(next.target())?;
    }
    Ok(())
}

fn augment_loaded_files<'a>(
    augmentor: &'a mut impl TemplateAugmentor,
    to_augment: &'a [LoadedFile],
) -> AppResult<Vec<AugmentedFile<'a>>> {
    to_augment
        .into_iter()
        .map(|next| {
            let augmented = augmentor.try_replace(next.content())?;
            Ok((augmented, next.path().as_path()))
        })
        .collect()
}

fn load_all_files(
    file_sys: &impl FileManipulator,
    to_load: impl IntoIterator<Item = FileToLoad>,
) -> AppResult<Vec<LoadedFile>> {
    to_load
        .into_iter()
        .map(|next_to_load| {
            let content = file_sys.read_file(next_to_load.source())?;
            Ok(LoadedFile::new(next_to_load.target().clone(), content))
        })
        .collect()
}

#[derive(Debug, new, Getters)]
struct LoadedFile {
    path: PathBuf,
    content: String,
}

fn determine_init_action(
    init_kind: InitKind,
    write_target: &Path,
    template_source: &Path,
    on_query_template: impl Fn(&Path) -> AppIoResult<Vec<NodeEntryMeta>>,
) -> AppIoResult<InitAction> {
    let nodes_write_target = on_query_template(write_target)?;
    return match init_kind {
        InitKind::OnlyEmpty => {
            if !nodes_write_target.is_empty() {
                Ok(InitAction::NotEmpty)
            } else {
                let templates =
                    all_templates_paths(template_source, write_target, on_query_template)?;
                Ok(InitAction::NoConflict(templates))
            }
        }
        InitKind::NoNameConflicts => {
            let templates = all_templates_paths(template_source, write_target, on_query_template)?;
            let target: Vec<NodeId> = templates
                .iter()
                .map(|next| (next.target_path(), next.node_type()))
                .collect();

            let nodes_write_target: Vec<NodeId> = many_node_metas_to_ids(&nodes_write_target);

            let to_return = match has_name_conflicts(&target, &nodes_write_target) {
                None => InitAction::NoConflict(templates),
                Some((conflict_path, _)) => InitAction::Conflict(conflict_path.to_path_buf()),
            };

            Ok(to_return)
        }
        InitKind::Override => {
            let templates = all_templates_paths(template_source, write_target, on_query_template)?;
            Ok(InitAction::NoConflict(templates))
        }
        InitKind::Purge => {
            let templates = all_templates_paths(template_source, write_target, on_query_template)?;
            Ok(InitAction::Purge(templates))
        }
    };

    fn all_templates_paths(
        template_source: &Path,
        write_target: &Path,
        on_query_template: impl Fn(&Path) -> AppIoResult<Vec<NodeEntryMeta>>,
    ) -> AppIoResult<Vec<SourceTargetNode>> {
        let templates = on_query_template(template_source)?;

        prepare_for_return(template_source, write_target, templates)
    }

    fn prepare_for_return(
        template_source: &Path,
        write_target: &Path,
        templates: Vec<NodeEntryMeta>,
    ) -> AppIoResult<Vec<SourceTargetNode>> {
        SourceTargetNode::opt_many_from_many_sources(template_source, write_target, templates)
            .ok_or_else(|| {
                AppIoError::custom("Could not get retrieve all paths from template files")
            })
    }
}

fn produce_set_from_paths<'a>(for_set: &'a [NodeId<'a>]) -> Option<HashSet<NodeId<'a>>> {
    let mut to_return: HashSet<NodeId<'a>> = Default::default();

    for to_insert in for_set {
        if !to_return.insert(*to_insert) {
            return None;
        }
    }
    Some(to_return)
}

fn has_name_conflicts<'a>(
    source: &'a [NodeId<'a>],
    target: &'a [NodeId<'a>],
) -> Option<NodeId<'a>> {
    let source_set = produce_set_from_paths(source)?;
    let target_set = produce_set_from_paths(target)?;
    let mut no_interection = source_set.intersection(&target_set);
    no_interection.next().copied()
}

fn many_node_metas_to_ids(many: &[NodeEntryMeta]) -> Vec<NodeId> {
    many.iter().map(|e| e.into()).collect()
}

#[cfg(test)]
mod testing {

    use crate::file_management::FileKind;

    use super::*;

    #[derive(Debug, Serialize, Deserialize)]
    struct TestInputDetermineAction {
        template_source: PathBuf,
        write_target: PathBuf,
        content_of_write_target: Vec<NodeEntryMeta>,
        content_of_template_folder: Vec<NodeEntryMeta>,
    }

    #[test]
    fn test_produce_set_from_uniques_paths() {
        let input: Vec<NodeEntryMeta> = from_ron_input_file!("list_to_set_unique_paths.ron");
        let input = many_node_metas_to_ids(&input);

        let mut actual: Vec<NodeId> = produce_set_from_paths(&input)
            .unwrap()
            .into_iter()
            .collect();

        actual.sort();
        let actual: Vec<NodeEntryMeta> = many_ids_to_meta_nodes(&actual);
        insta::assert_ron_snapshot!(actual);
    }
    #[test]
    fn test_none_for_unique_key() {
        let input: Vec<NodeEntryMeta> = from_ron_input_file!("list_to_set_not_unique_paths.ron");
        let input = many_node_metas_to_ids(&input);
        let actual = produce_set_from_paths(&input);
        assert!(actual.is_none());
    }

    #[test]
    fn test_detect_no_paths_intersection() {
        let (left, right): (Vec<NodeEntryMeta>, Vec<NodeEntryMeta>) =
            from_ron_input_file!("interection_unique.ron");
        let (left, right) = (
            many_node_metas_to_ids(&left),
            many_node_metas_to_ids(&right),
        );
        let actual = has_name_conflicts(&left, &right);
        assert!(actual.is_none());
    }
    #[test]
    fn test_detect_paths_intersection() {
        let (left, right): (Vec<NodeEntryMeta>, Vec<NodeEntryMeta>) =
            from_ron_input_file!("interection_not_unique.ron");
        let (left, right) = (
            many_node_metas_to_ids(&left),
            many_node_metas_to_ids(&right),
        );
        let actual = has_name_conflicts(&left, &right).unwrap();
        let expected = NodeEntryMeta::new(FileKind::File, PathBuf::from("/some/target/z"));
        assert_eq!(expected, single_id_to_node_entry_meta(actual));
    }

    #[test]
    fn test_detect_name_conflict() {
        let init_kind = InitKind::NoNameConflicts;
        let input: TestInputDetermineAction = from_ron_input_file!("with_name_conflicts.ron");
        let callback = callback_on_query_content(&input);
        let actual = determine_init_action(
            init_kind,
            &input.write_target,
            &input.template_source,
            callback,
        )
        .unwrap();
        insta::assert_ron_snapshot!(actual);
    }
    #[test]
    fn test_return_without_name_conflict() {
        let init_kind = InitKind::NoNameConflicts;
        let input: TestInputDetermineAction = from_ron_input_file!("without_name_conflicts.ron");
        let callback = callback_on_query_content(&input);
        let actual = determine_init_action(
            init_kind,
            &input.write_target,
            &input.template_source,
            callback,
        )
        .unwrap();
        insta::assert_ron_snapshot!(actual);
    }

    #[test]
    fn test_err_return_not_empty() {
        let init_kind = InitKind::OnlyEmpty;
        let input: TestInputDetermineAction = from_ron_input_file!("err_return_not_empty.ron");
        let actual = determine_init_action(
            init_kind,
            &input.write_target,
            Path::new("/a/b"),
            |actual_path| {
                if actual_path == input.write_target {
                    Ok(input.content_of_write_target.clone())
                } else {
                    panic!("Should not ask for all paths within the template source if {} is set and write target is not empty", init_kind)
                }
            },
        );
        assert!(matches!(actual, Ok(InitAction::NotEmpty)));
    }

    #[test]
    fn test_return_all_templates_for_empty_target() {
        let init_kind = InitKind::OnlyEmpty;
        let input: TestInputDetermineAction = from_ron_input_file!("test_return_if_empty.ron");
        let callback = callback_on_query_content(&input);
        let actual = determine_init_action(
            init_kind,
            &input.write_target,
            &input.template_source,
            callback,
        )
        .unwrap();
        insta::assert_ron_snapshot!(actual);
    }

    #[test]
    fn test_return_for_purge() {
        let init_kind = InitKind::Purge;
        let input: TestInputDetermineAction = from_ron_input_file!("with_name_conflicts.ron");
        let callback = callback_on_query_content(&input);
        let actual = determine_init_action(
            init_kind,
            &input.write_target,
            &input.template_source,
            callback,
        )
        .unwrap();
        insta::assert_ron_snapshot!(actual);
    }

    #[test]
    fn test_return_for_override() {
        let init_kind = InitKind::Override;
        let input: TestInputDetermineAction = from_ron_input_file!("with_name_conflicts.ron");
        let callback = callback_on_query_content(&input);
        let actual = determine_init_action(
            init_kind,
            &input.write_target,
            &input.template_source,
            callback,
        )
        .unwrap();
        insta::assert_ron_snapshot!(actual);
    }

    fn callback_on_query_content(
        input: &TestInputDetermineAction,
    ) -> impl Fn(&Path) -> AppIoResult<Vec<NodeEntryMeta>> + '_ {
        |actual_path| {
            if actual_path == input.write_target {
                Ok(input.content_of_write_target.clone())
            } else if actual_path == input.template_source {
                Ok(input.content_of_template_folder.clone())
            } else {
                panic!("Actual path {:?} is not valid", actual_path);
            }
        }
    }

    fn single_id_to_node_entry_meta(id: NodeId) -> NodeEntryMeta {
        NodeEntryMeta::new(id.1, id.0.to_path_buf())
    }

    fn many_ids_to_meta_nodes(many: &[NodeId]) -> Vec<NodeEntryMeta> {
        many.iter()
            .map(|&e| single_id_to_node_entry_meta(e))
            .collect()
    }
}
