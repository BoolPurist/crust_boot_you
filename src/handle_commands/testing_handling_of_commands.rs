use super::*;
use pretty_assertions::assert_eq;
use std::{io, path::PathBuf};

use crate::app_traits::path_provider::{MockPathProvider, TestPathProvider};
use crate::{app_traits::file_manipulator::MockFileManipulator, cli::InitKind};

fn setup_save_until_template_check(
    path_provider: &mut MockPathProvider,
    file_manipulator: &mut MockFileManipulator,
    name: &NotEmptyText,
    template_path: &Path,
    exits: bool,
) -> PathBuf {
    let expected_name = name.clone();
    let expected_name_for_files = name.clone();
    let expected_template_path = template_path.to_path_buf();
    let expected_template_path_for_files =
        template_path.to_path_buf().join(constants::FILES_FOLDER);

    let return_value = expected_template_path_for_files.to_path_buf();
    let expected_path_to_check = expected_template_path.clone();
    path_provider
        .expect_specific_entry_template()
        .times(1)
        .withf(move |actual_name| *actual_name == expected_name)
        .return_once(|_| Ok(expected_template_path));
    path_provider
        .expect_specific_entry_template_files()
        .times(1)
        .withf(move |actual_name| *actual_name == expected_name_for_files)
        .return_once(|_| Ok(expected_template_path_for_files));

    file_manipulator
        .expect_try_exits()
        .withf(move |actual_path_to_check| actual_path_to_check == expected_path_to_check)
        .return_once(move |_| Ok(exits));

    return_value
}

fn setup_ensure_dir(file_manipulator: &mut MockFileManipulator, expected_target_file_path: &Path) {
    let expected = expected_target_file_path.to_path_buf();
    file_manipulator
        .expect_ensure_dir()
        .times(1)
        .withf(move |param| param == expected)
        .return_once(|_| Ok(()));
}

#[ignore]
#[test]
fn deny_creation_of_template_if_already_there() {
    let mut path_provider = MockPathProvider::default();
    let mut file_manipulator = MockFileManipulator::default();
    let on_file_detect = |_: &_| {
        panic!("Should not come this far");
    };
    let given_name = NotEmptyText::new("Not there".to_string()).unwrap();
    let given_path = PathBuf::from("some_template_name");

    let expected_name = given_name.clone();
    let expected_template_path = PathBuf::from("some/template_path").join(expected_name.as_ref());

    setup_save_until_template_check(
        &mut path_provider,
        &mut file_manipulator,
        &given_name,
        &expected_template_path,
        true,
    );

    let message = save_template(
        path_provider,
        file_manipulator,
        on_file_detect,
        &given_name,
        given_path,
    )
    .unwrap_err();
    let expected_msg = save_err_already_created_template(&given_name);

    assert_eq!(expected_msg, message.to_string());
}

#[ignore]
#[test]
fn create_template_from_one_file() {
    let mut path_provider = MockPathProvider::default();
    let mut file_manipulator = MockFileManipulator::default();
    let on_file_detect = |_: &_| Ok(FileKind::File);
    let given_name = NotEmptyText::new("Not there".to_string()).unwrap();
    let given_path = PathBuf::from("some_template_name");

    let expected_template_path = setup_save_until_template_check(
        &mut path_provider,
        &mut file_manipulator,
        &given_name,
        &given_path,
        false,
    );

    let expected_target_file_path = expected_template_path
        .join(constants::FILES_FOLDER)
        .join(given_name.as_ref());
    setup_ensure_dir(&mut file_manipulator, &expected_template_path);

    let expected_source_path = given_path.clone();
    file_manipulator
        .expect_copy_file()
        .withf(move |actual_source_path, actual_target_path| {
            actual_source_path == expected_source_path
                && actual_target_path == expected_target_file_path
        })
        .returning(|_, _| Ok(()));

    let output = save_template(
        path_provider,
        file_manipulator,
        on_file_detect,
        &given_name,
        &given_path,
    )
    .unwrap();

    let expected_msg = success_save_msg(&given_name, "file", &given_path);
    assert_eq!(expected_msg, output);
}
#[test]
fn create_template_from_folder() {
    let mut path_provider = MockPathProvider::default();
    let mut file_manipulator = MockFileManipulator::default();
    let on_file_detect = |_: &_| Ok(FileKind::Folder);
    let given_name = NotEmptyText::new("I am a folder/ hey".to_string()).unwrap();
    let given_path = PathBuf::from("some_template_name");

    let expected_template_path = setup_save_until_template_check(
        &mut path_provider,
        &mut file_manipulator,
        &given_name,
        &given_path,
        false,
    );

    setup_ensure_dir(&mut file_manipulator, &expected_template_path);

    let expected_source_path = given_path.clone();
    let expected_target_path = expected_template_path.clone();
    file_manipulator
        .expect_copy_dir()
        .times(1)
        .withf(move |actual_source_path, actual_target_path| {
            actual_source_path == expected_source_path && actual_target_path == expected_target_path
        })
        .return_once(|_, _| Ok(()));

    let output = save_template(
        path_provider,
        file_manipulator,
        on_file_detect,
        &given_name,
        &given_path,
    )
    .unwrap();

    let expected_error_message = success_save_msg(&given_name, "folder", &given_path);
    assert_eq!(expected_error_message, output);
}
#[test]
fn may_not_manipulate_files_if_no_folder_or_file() {
    let name: NotEmptyText = NotEmptyText::new("Some_Name".to_owned()).unwrap();
    let on_detect_file_kind = |_: &_| Ok(FileKind::Symlink);
    let source_path = PathBuf::from("some/source");
    let mut path_fetcher = MockPathProvider::default();
    let mut file_manipulator = MockFileManipulator::default();

    setup_save_until_template_check(
        &mut path_fetcher,
        &mut file_manipulator,
        &name,
        &source_path,
        false,
    );

    save_template(
        path_fetcher,
        file_manipulator,
        on_detect_file_kind,
        &name,
        &source_path,
    )
    .unwrap_err();
}

#[test]
fn load_err_bail_for_source_path_not_querable() {
    let data_folder = PathBuf::from("/some/data");
    let paths = TestPathProvider::new(
        data_folder.clone(),
        data_folder.clone(),
        data_folder.clone(),
    );
    let name = NotEmptyText::new("I am not ther".to_string()).unwrap();
    let assumed_template_entry_path = paths.specific_entry_template_files(&name).unwrap();
    let mut files = MockFileManipulator::default();

    files
        .expect_try_exits()
        .times(1)
        .withf(move |to_check| to_check == assumed_template_entry_path)
        .returning(|_| Err(AppIoError::custom("a")));
    let args = LoadTemplateArg::new(name, InitKind::default());
    handle_load_template(paths, files, &args).unwrap_err();
}

#[test]
fn load_err_bail_for_source_path_does_not_exits() {
    let data_folder = PathBuf::from("/some/data");
    let paths = TestPathProvider::new(
        data_folder.clone(),
        data_folder.clone(),
        data_folder.clone(),
    );
    let name = NotEmptyText::new("I am not ther".to_string()).unwrap();
    let mut files = MockFileManipulator::default();
    let assumed_template_entry_path = paths.specific_entry_template_files(&name).unwrap();

    files
        .expect_try_exits()
        .times(1)
        .withf(move |to_check| to_check == assumed_template_entry_path)
        .returning(|_| Ok(false));
    let args = LoadTemplateArg::new(name, InitKind::default());
    handle_load_template(paths, files, &args).unwrap_err();
}

#[test]
fn load_success_copy_template_to_source_path() {
    let data_folder = PathBuf::from("/some/data");
    let cwd = PathBuf::from("/coding/rust");
    let paths = TestPathProvider::new(data_folder.clone(), data_folder.clone(), cwd.clone());
    let name = NotEmptyText::new("AAA".to_string()).unwrap();
    let assumed_template_entry_path = paths.specific_entry_template_files(&name).unwrap();
    let expected_ensured_template_folder = assumed_template_entry_path.clone();

    // Mocking
    let mut files = MockFileManipulator::default();
    files
        .expect_try_exits()
        .times(1)
        .withf(move |to_check| to_check == assumed_template_entry_path)
        .returning(|_| Ok(true));
    let expected_target_location = cwd.clone();
    files
        .expect_copy_dir()
        .times(1)
        .withf(move |actual_source_path, actual_target_path| {
            actual_source_path == expected_ensured_template_folder
                && actual_target_path == expected_target_location
        })
        .returning(|_, _| Ok(()));

    // Act
    let args = LoadTemplateArg::new(name.clone(), InitKind::default());
    let output = handle_load_template(paths, files, &args).unwrap();

    // Assert
    let expected_output = format!(
        "Folder {:?} filled with content from Template ({})",
        cwd, name
    );
    assert_eq!(expected_output, output);
}

#[test]
fn list_should_return_all_templates() {
    let mut paths = MockPathProvider::default();
    let expected_template_entry = PathBuf::from("some/all_templates");
    paths
        .expect_general_template_entry()
        .times(1)
        .returning(|| Ok(PathBuf::from("some/all_templates")));
    let mut files = MockFileManipulator::default();

    let expected_ensure_template_entry = expected_template_entry.clone();
    files
        .expect_ensure_dir()
        .times(1)
        .withf(move |actual_template_entry| actual_template_entry == expected_ensure_template_entry)
        .returning(|_| Ok(()));
    files
        .expect_list_first_level_dir()
        .times(1)
        .withf(move |actual_listing_path| actual_listing_path == expected_template_entry)
        .returning(|_| Ok(vec![PathBuf::from("rust"), PathBuf::from("python")]));

    let output = handle_list_template(paths, files).unwrap();
    let expected_ouput = format!("{}\nrust\npython\n", constants::TITLE_LIST_RESULT);
    assert_eq!(expected_ouput, output);
}

fn setup_act_assert_delete(
    expected_delete_return: AppIoResult,
    given_name: NotEmptyText,
) -> ReturnToUser {
    let mut path_provider = MockPathProvider::default();

    let mut file_manipulator = MockFileManipulator::default();

    let expected_template_path = PathBuf::from("/some/data").join(given_name.as_ref());
    let expected_target_path = expected_template_path.clone();
    let expected_name = given_name.clone();
    path_provider
        .expect_specific_entry_template()
        .times(1)
        .withf(move |actual_name| *actual_name == expected_name)
        .return_once(|_| Ok(expected_template_path));
    file_manipulator
        .expect_delete_whole_folder()
        .withf(move |actual_target_folder| actual_target_folder == expected_target_path)
        .return_once(move |_| expected_delete_return);
    handle_delete_template(path_provider, file_manipulator, &given_name)
}

#[test]
fn delete_template() {
    let given_name = NotEmptyText::new("to_delete".to_string()).unwrap();
    let expected_message = format!("Template ({}) was deleted.", given_name.as_ref());
    let message = setup_act_assert_delete(Ok(()), given_name).unwrap();
    assert_eq!(expected_message, message);
}
#[test]
fn delete_err_if_no_template() {
    let given_name = NotEmptyText::new("not_there".to_string()).unwrap();
    let expected_message = error_delete_msg_not_found(&given_name);
    let message = setup_act_assert_delete(Err(AppIoError::NotFound), given_name).unwrap_err();
    assert_eq!(expected_message, message.to_string());
}
#[test]
fn delete_some_other_err() {
    let given_name = NotEmptyText::new("not_there".to_string()).unwrap();

    let error = io::Error::new(io::ErrorKind::Other, "some error").into();
    let expected_message = error_delet_msg_other_err(&given_name, error);
    let message = setup_act_assert_delete(
        Err(io::Error::new(io::ErrorKind::Other, "some error").into()),
        given_name,
    )
    .unwrap_err();
    assert_eq!(expected_message, message.to_string());
}
