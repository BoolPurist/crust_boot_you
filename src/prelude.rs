use std::io;

pub use crate::constants;
pub use anyhow::Context;

pub type IoResult<T = ()> = Result<T, io::Error>;
pub type AppError = anyhow::Error;
pub type AppResult<T = ()> = Result<T, AppError>;

pub type ReturnToUser = AppResult<String>;
pub type PathResult = AppResult<PathBuf>;
pub type AppIoResult<T = ()> = Result<T, AppIoError>;

pub use crate::app_traits::file_manipulator::FileManipulator;
pub use crate::app_traits::path_provider::PathProvider;
pub use crate::file_management::AppIoError;
pub use crate::NotEmptyText;
pub use std::path::{Path, PathBuf};

#[cfg(test)]
pub use mockall::automock;
