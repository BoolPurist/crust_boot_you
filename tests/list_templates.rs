use common::prelude::*;
mod common;

use crust_boot_you::handle_commands;

#[named]
#[test]
fn list_avaiable_templates() {
    let setup = TestSetup::only_actual(actual!());

    let test_paths = TestPathProvider::clone_from(setup.path_to_temp(), "data", "config", "cwd");

    let result = handle_commands::handle_list_template(&test_paths, setup.os_mani()).unwrap();
    setup.assert_with_expected();

    insta::with_settings!({filters => vec![
        (r"/tmp/\.[\w]+/", "/[TMP_ROOT]/"),
    ]},
    {
        insta::assert_display_snapshot!(result);
    });
}

#[named]
#[test]
fn list_template_no_template() {
    let setup = TestSetup::only_actual(actual!());

    let test_paths = TestPathProvider::clone_from(setup.path_to_temp(), "data", "config", "cwd");

    let result = handle_commands::handle_list_template(&test_paths, setup.os_mani()).unwrap();
    setup.assert_with_expected();
    insta::assert_display_snapshot!(result);
}
