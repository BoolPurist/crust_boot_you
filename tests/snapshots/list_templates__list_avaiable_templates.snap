---
source: tests/list_templates.rs
expression: result
---
List of all templates:
b  "/[TMP_ROOT]/data/templates/b"
a  "/[TMP_ROOT]/data/templates/a"

